remill/BC/DeadStoreEliminator.cpp

```diff
@@ -0,0 +1,532 @@
+    if (ptr == offset_map.end()) {
+      return VisitResult::NoProgress;
+    } else {
+      // update value
+      offset_map[&I] = ptr->second;
+      LOG(INFO) << "offsetting: " << LLVMThingToString(&I);
+      return VisitResult::Progress;
+    }
+  }
+}
+
+// Visit an add instruction and update the offset map.
+VisitResult ForwardAliasVisitor::visitAdd(llvm::BinaryOperator &I) {
+  LOG(INFO) << "Entered add instruction";
+  return ForwardAliasVisitor::visitBinaryOp_(I, true);
+}
+
+// Visit a sub instruction and update the offset map.
+VisitResult ForwardAliasVisitor::visitSub(llvm::BinaryOperator &I) {
+  LOG(INFO) << "Entered sub instruction";
+  return ForwardAliasVisitor::visitBinaryOp_(I, false);
+}
+
+// Visit an add or sub instruction.
+VisitResult ForwardAliasVisitor::visitBinaryOp_(llvm::BinaryOperator &I, bool plus) {
+  auto val1 = I.getOperand(0);
+  auto val2 = I.getOperand(1);
+  if (exclude.count(val1) || exclude.count(val2)) {
+    exclude.insert(&I);
+    LOG(INFO) << "excluding: " << LLVMThingToString(&I);
+    return VisitResult::Progress;
+  } else {
+    if (auto cint = llvm::dyn_cast<llvm::ConstantInt>(val1)) {
+      auto ptr = offset_map.find(val2);
+      if (ptr == offset_map.end()) {
+        return VisitResult::NoProgress;
+      } else {
+        auto offset = (plus ? ptr->second + cint->getZExtValue() : ptr->second - cint->getZExtValue());
+        offset_map.insert({&I, offset});
+        LOG(INFO) << "offsetting: " << LLVMThingToString(&I);
+        return VisitResult::Progress;
+      }
+    } else if (auto cint = llvm::dyn_cast<llvm::ConstantInt>(val2)) {
+      auto ptr = offset_map.find(val1);
+      if (ptr == offset_map.end()) {
+        return VisitResult::NoProgress;
+      } else {
+        auto offset = (plus ? ptr->second + cint->getZExtValue() : ptr->second - cint->getZExtValue());
+        offset_map.insert({&I, offset});
+        LOG(INFO) << "offsetting: " << LLVMThingToString(&I);
+        return VisitResult::Progress;
+      }
+    } else {
+      // check if both are in the offset_map
+      auto ptr1 = offset_map.find(val1);
+      auto ptr2 = offset_map.find(val2);
+      if (ptr1 == offset_map.end() || ptr2 == offset_map.end()) {
+        return VisitResult::NoProgress;
+      } else {
+        auto offset = (plus ? ptr1->second + ptr2->second : ptr1->second - ptr2->second);
+        offset_map.insert({&I, offset});
+        LOG(INFO) << "offsetting: " << LLVMThingToString(&I);
+        return VisitResult::Progress;
+      }
+    }
+  }
+}
+
+// Visit a PHI node and update the offset map.
+VisitResult ForwardAliasVisitor::visitPHINode(llvm::PHINode &I) {
+  LOG(INFO) << "Entered PHI node";
+  // iterate over each operand
+  auto in_offset_map = false;
+  uint64_t offset;
+  for (auto &operand : I.operands()) {
+    if (exclude.count(operand)) {
+      // fail if some operands are excluded and others are state offsets
+      if (in_offset_map) {
+        return VisitResult::Error;
```

Comment:
This check isn't sufficient to check the property of interest. I think what you want is to have a `bool` that is something like `bool in_exclude_set = false;` then you set to `true` if you find something in the exclude set, but keep going! (just like with `in_offset_map`). At the end of the function, `if (in_offset_map && in_exclude_set)` then you want to error out.

The key is that every time we visit the phi, we want to inspect all operands, so we can never return early inside this loop.